### ISSUE TITLE
Modular_Station=0: Wait for Data_Exchange-ready won't proceed to Data Exchange stage

### ISSUE THREAD

**Issue by marcosgatcomputer** [mbuesch/pyprofibus#7]
Hello,
I have an issue I have not been able to solve that I'd like to share.

I have ported the Profibus Slave originally intended for the [Texas Instruments Industrial Communication Development Boards](http://www.ti.com/tool/TMDSICE3359) to run on a Beaglebone Black (BBB) SBC. This is, as far as I know, the most affordable Profibus slave one can get and, in that economic sense a perfect match for *pyprofibus*.

Unfortunately, I have not been able to make them talk together. The slave works fine out of the box with the default GSD file using a PLC as a master but when I wire it to my laptop and try to connect with *pyprofibus* it gets stuck in a loop going from waiting for data_exchange-ready to the initial state instead of proceeding to the data exchange stage.

This is the log I got:
```
DPM1: Trying to initialize slave 2...
PHY-serial: TX   10 02 63 49 AE 16
PHY-serial: RX   10 63 02 00 65 16
DPM1: slave[02].state --> 'wait for diag'
DPM1: Requesting Slave_Diag from slave 2...
PHY-serial: TX   68 05 05 68 82 E3 6D 3C 3E 4C 16
PHY-serial: RX   68 0B 0B 68 E3 82 08 3E 3C 02 05 00 FF 0C DA D3 16
DPM1: slave[02].state --> 'wait for Prm'
DPM1: Sending Set_Prm to slave 2...
PHY-serial: TX   68 0F 0F 68 82 E3 5D 3D 3E B8 05 01 00 0C DA 01 40 00 00 22 16
PHY-serial: RX   E5
DPM1: slave[02].state --> 'wait for Cfg'
DPM1: Sending Chk_Cfg to slave 2...
PHY-serial: TX   68 05 05 68 82 E3 7D 3E 3E 5E 16
PHY-serial: RX   E5
DPM1: slave[02].state --> 'wait for Data_Exchange-ready'
DPM1: Requesting Slave_Diag (WDXRDY) from slave 2...
PHY-serial: TX   68 05 05 68 82 E3 5D 3C 3E 3C 16
PHY-serial: RX   68 0B 0B 68 E3 82 08 3E 3C 02 0C 00 63 0C DA 3E 16
PHY-serial: TX   68 05 05 68 82 E3 7D 3C 3E 5C 16
PHY-serial: RX   68 0B 0B 68 E3 82 08 3E 3C 02 05 00 FF 0C DA D3 16
DPM1: slave[02].state --> 'init'
DPM1: Trying to initialize slave 2...

...and repeats in a loop...
```
As you can see, the problem is the slave answers to the diagnostics telegram raising the `Stat_Diag` bit and in the second attempt it comes back with the master address reset at `0xFF` indicating its configuration needs to be updated (although the master address was correct in the first telegram, `0x63`).

I've tried to change the configuration every possible way but after a couple of days, I think my next shot will be to use a sniffer to monitor the connection to a PLC master. In the meantime maybe there is something obvious I'm missing... The slave side is a library and debugging it is quite a slow process (I mean trying to find the mismatch in the configuration from the slave side).

Some additional info on my setup:

-[TI Profibus Slave Demo Wiki page](http://processors.wiki.ti.com/index.php/PROFIBUS_DP_slave_demo_setup_on_AM335x). I don't have Profitrace but I assume my setup should work exactly the same as TI's ICCS board.

-Python, conf and GSD files can be found in this [repo](https://github.com/marcosgatcomputer/BBB_Profibus_Slave/tree/master/pyprofibus). I will be uploading all binaries and source there soon.

Thank you.

**Comment by mbuesch** [comment]
Thanks for your report.

I can't really tell why this does not work. Pyprofibus does not correctly implement all aspects of the Profibus protocol, yet. So this most likely is due to a bug or a missing feature in pyprofibus.
One thing that is not implemented correctly is the bus timing requirements. I fixed some of these things in the latest commits, but we are still far away from having compliant timing.

I would suggest you play around with the DP state machine code and modify it until you get different results.

**Comment by marcosgatcomputer** [comment]
Thank you for your comments. 

I will think about it and try to make more tests over the week and post back my findings next week.

**Comment by marcosgatcomputer** [comment]
Hello again,

I finally had some time to check why pyprofibus was not working with my BBB slave.

As it turns out I had two issues, none of them related to bugs or pyprofibus itself but to the configuration I was trying to use.

First off, the GSD that came with the Profibus stack was defined as non-modular (`Modular_Station      = 0`). I'm not really sure you can define a non-modular slave, as far as I know, the only way to set the data sizes is with a module, and that's what they did here too with a subsequent line:

`Module = "8 Byte Out, 8 Byte In " 0x17,0x27`

Maybe there is a different way to do it. I was under the (wrong) impression that if you define the slave as non-modular, pyprofibus was considering this single module somehow. Anyway, that was pretty easy to solve just changing the GSD to modular (`Modular_Station = 1`) and adding the module in pyprofibus conf file:

`module_0=8 Byte Out, 8 Byte In`

Then I had a problem with the output buffer size. I was trying to run with an 8-byte output (as defined by the 0x27 code in the configuration):

`outData[slaveDesc.slaveAddr] = bytearray([0x00]*8)`

But I'm afraid I must have misunderstood what this buffer is because if I define it like that (8 bytes long) the state machine gets stuck in the "wait for data exchange" loop. What I find surprising is that any dimension smaller than 8 works fine whilst anything bigger gives the obvious `IndexError: list index out of range` while trying to flip outputs to inputs.

For reference, this is the log I get when I have a successful data exchange:

```
DPM1: Trying to initialize slave 2...
PHY-serial: TX   10 02 63 49 AE 16
PHY-serial: TX   10 02 63 49 AE 16
PHY-serial: TX   10 02 63 49 AE 16
PHY-serial: TX   10 02 63 49 AE 16
PHY-serial: RX   10 63 02 00 65 16
DPM1: slave[02].state --> 'wait for diag'
DPM1: Requesting Slave_Diag from slave 2...
PHY-serial: TX   68 05 05 68 82 E3 6D 3C 3E 4C 16
PHY-serial: RX   68 0B 0B 68 E3 82 08 3E 3C 02 05 00 FF 0C DA D3 16
DPM1: slave[02].state --> 'wait for Prm'
DPM1: Sending Set_Prm to slave 2...
PHY-serial: TX   68 0F 0F 68 82 E3 5D 3D 3E B8 0A 01 00 0C DA 01 00 00 00 E7 16
PHY-serial: RX   E5
DPM1: slave[02].state --> 'wait for Cfg'
DPM1: Sending Chk_Cfg to slave 2...
PHY-serial: TX   68 07 07 68 82 E3 7D 3E 3E 17 27 9C 16
PHY-serial: RX   E5
DPM1: slave[02].state --> 'wait for Data_Exchange-ready'
DPM1: Requesting Slave_Diag (WDXRDY) from slave 2...
PHY-serial: TX   68 05 05 68 82 E3 5D 3C 3E 3C 16
PHY-serial: RX   68 0B 0B 68 E3 82 08 3E 3C 00 0C 00 63 0C DA 3C 16
DPM1: slave[02].state --> 'Data_Exchange'
DPM1: Initialization finished. Running Data_Exchange with slave 2...
PHY-serial: TX   68 05 05 68 02 63 7D 00 00 E2 16
PHY-serial: RX   68 0B 0B 68 63 02 08 02 02 02 02 02 02 02 02 7D 16
PHY-serial: TX   68 05 05 68 02 63 5D 02 00 C4 16
PHY-serial: RX   10 63 02 03 68 16
Terminating: Service not active on slave 2

```
I'm still not sure why the process is terminated with the message on the last line after just one data exchange.

I'm closing the issue, but your comments on my musings would be more than welcome.

Thank you.

**Comment by mbuesch** [comment]
Thanks for your great analysis.

I re-open this issue, because I do actually think that the handling of Modular_Station in pyprofibus is wrong.
According to the DP spec it seems that even for "compact" (= non-modular) devices one Module is the way to specify the configuration and parameter bytes.
So we should probably allow one (only one?) Module if Modular_Station is 0.

If the out-buffer size of 8 bytes does not work with your module that has 8 bytes out, there probably is something else going wrong. Maybe something else in the GSD parser/interpreter or some machinery related to that.
I think that's worth investigating.

**Comment by marcosgatcomputer** [comment]
Hello again,

Thanks for reacting so quickly.

I think it what you say about the non-modular slaves makes sense. It should be very easy to change the parser to take a single module and update its configuration. The workaround was even easier though so i took the lazy approach.

Regarding the buffer size, you are right too. There must be something wrong but I still could not find where (probably on my slave's code). 

What I did is change the configuration in the slave to make the input and output buffers 32 bytes instead with:

`Module = "32 Byte Input, 32 Byte Output" 0x5F,0x6F`

This defines the buffers as word size instead of byte size and 32 bytes instead of 8. With this change everything works as intended, I go into the data exchange cycle and stay there until I cancel. I can see the inputs and output changing too. So, mission accomplished! 

I'm going to dig into why this is happening. Last week I was doing some tests with a SIEMENS ET200S and I was witnessing something very similar: the data exchange would end after just one iteration.

I will report back later in the week if I manage to find out something useful.

Thank you again for your support.

**Comment by mbuesch** [comment]
Is it possible that you send me your original GSD file that contains Modular_Station = 0?

**Comment by marcosgatcomputer** [comment]
You can find it [here](https://github.com/marcosgatcomputer/BBB_Profibus_Slave/tree/master/pyprofibus).

**Comment by marcosgatcomputer** [comment]
Hello,

I have been looking into the buffer sizes and I still don't get it.

As I said above, if I define the data blocks of size 16, no matter if I use words or bytes, everything works fine. For instance, I can define:

`Module = "16 Word Input, 16 Byte Output" 0x5F,0x6F`

For 16 words input-output (as long as I define the buffers 32 bytes on the slave and the output_data on pyprofibus). Or:

`Module = "16 Byte Input, 16 Byte Output" 0x1F,0x2F`

But if I try to use 8 bytes:

`Module = "8 Byte Input, 8 Byte Output" 0x17,0x27`

The slave reports a parametrization error when I try to run it with an 8-byte buffer or it stops after the first data exchange with the `Terminating: Service not active` error if the buffer is smaller.

This is not a big deal, of course, I can use 32-byte data blocks but it's kind of annoying not knowing what is wrong.

This is similar to what I witnessed last week with the ET200S slave. I was using the following configuration:

```
module_0=1st general ID (0x00)
module_1=2nd general ID (0x00)
module_2=3rd general ID (0x00)
module_3=Master_I Slave_Q  8 Wo unit (0x57)
module_4=Master_I Slave_Q 16 Wo unit (0x5F)
module_5=Master_I Slave_Q 16 Wo unit (0x5F)
module_6=Master_I Slave_Q 16 Wo unit (0x5F)
module_7=Master_Q Slave_I  8 Wo unit (0x67)
module_8=Master_Q Slave_I  2 Wo unit (0x61)
module_9=Master_I Slave_Q  2 Wo unit (0x51)
```
To replicate what I had on the slave's side.  And the transaction ended like this:

```
PHY-serial: TX   68 17 17 68 08 01 7D 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 86 16
PHY-serial: RX   10 01 08 03 0C 16
Terminating: Service not active on slave 8
```

As you can see there are 20 placeholder bytes (0x00) for the 10 words that the slave should be sending back.

I will take a look at the code to be sure, but this looks like it might be a pyprofibus bug for module data blocks of size 8. Or maybe I'm missing something else...

**Comment by mbuesch** [comment]
Yes, there certainly might be a bug.
However it's hard to say what's going on.
The `Service not active` error happens, if the slave terminates it's service. So something unexpected (too much or not enough) data has been sent to the slave.

For the Nonmodular-Device issue I will try the following:
If Modular_Station = 0, then automatically use all Modules that are present in the GSD file.

That should be easy and straightforward to implement.
I'll do that and send it to you for testing, if you're interested.

**Comment by marcosgatcomputer** [comment]
OK, thanks. I'll do some more testing tomorrow and get back if I learn something useful.

I agree with your approach to fix the non-modular slave. I can test it, no problem, and if you happen to have or are able to get a BBB I can share the binaries in case you want to check it out. I will get busy writing the instructions to modify TI's PDK very soon.

**Comment by mbuesch** [comment]
I added a possible fix for Modular_Station=0 to the branch `gsd_compact_slave`:
https://github.com/mbuesch/pyprofibus/tree/gsd_compact_slave

**Comment by marcosgatcomputer** [comment]
Hello,

The parser works out of the box now for the non-modular slave. Thank you for fixing it.

If you agree, you can close the issue. I'll open a new issue if I ever find out what's going on with the 8-byte data blocks.

**Comment by mbuesch** [comment]
Thanks.

This has been fixed in cd28982bc6a29a359bda7842ade4d83e784b6470
Release: pyprofibus-1.6

For the data block problem please open a second issue.
